src/python/pants/jvm/resolve/coursier_fetch_integration_test.py

```diff
@@ -668,3 +668,60 @@ def test_transitive_excludes(rule_runner: RuleRunner) -> None:
     entries = resolve.entries
     assert any(i for i in entries if i.coord.artifact == "jackson-databind")
     assert not any(i for i in entries if i.coord.artifact == "jackson-core")
+
+
+@pytest.mark.xfail(
+    reason="coursier bug: https://github.com/coursier/coursier/issues/2884", strict=True
+)
+@maybe_skip_jdk_test
+def test_missing_entry_for_transitive_dependency(rule_runner: RuleRunner) -> None:
+    resolve = rule_runner.request(
+        CoursierResolvedLockfile,
+        [
+            ArtifactRequirements(
+                [
+                    Coordinate(
+                        group="org.apache.hive",
+                        artifact="hive-exec",
+                        version="1.1.0",
+                    )
+                    .as_requirement()
+                    .with_extra_excludes(
+                        "org.apache.calcite:calcite-avatica",
+                        "org.apache.calcite:calcite-core",
+                        "jdk.tools:jdk.tools",
+                    )
+                ]
+            )
+        ],
+    )
+
+    lookup = {(entry.coord.group, entry.coord.artifact): entry for entry in resolve.entries}
+    missing = []
+    for entry in resolve.entries:
+        for d in entry.dependencies:
+            coord = (d.group, d.artifact)
+            if coord not in lookup:
+                missing.append(coord)
+
+    # We expect all the dependencies to have an entry, but right now it's not true
+    # for ("junit", "junit") and ("org.apache.curator", "apache-curator").
+    # TODO Remove the workaround once the bug is fixed.
+    assert missing == []
```

Comment:
If we're confirming whether the workaround needs to remain, I think it would be better to write this as a "positive" test:

```suggestion
    assert missing == [("junit", "junit"), ("org.apache.curator", "apache-curator")]
```

And remove the `xfail` (and potentially update the comment to be explicit that we might hope that it's _not_ empty).

In particular, the `xfail` is currently quite broad: _any_ exception will cause the `xfail` to "pass". For instance, if the underlying bug fix was removed, I suspect the `rule_runner.request(...)` would propagate the `KeyError` that was originally occurring, and thus cause the test to "pass".

If we instead test the explicit violation of our expectations here, we at least confirm the `rule_runner.request(...)` call works i.e. that the original #20162 bug remains fixed.